Lazy load user capabilities in WP_User object

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/58001

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Test

spacedmonkey committed:
Add wp_lazyload_user_meta function and use it.

spacedmonkey committed:
Fix tests.

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey committed:
Reduce number of expected queries in cacheResults test

The number of expected queries in the cacheResults test has been decreased from two to one. Specifically, the user data query is no longer expected, only the user meta data query remains. The test function has been updated accordingly to reflect this change.

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -61,7 +61,7 @@ class WP_User {
 	 *             and boolean values represent whether the user has that capability.
 	 */
-	public $caps = array();
+	protected $caps = array();
```
Why is this changed to `protected`? Wouldn't it be better to make it `private` instead?

spacedmonkey replied:
I don't think we can make this private. If someone is extending the WP_User class, then this property will then become unaccessible. There is no need for it to be private, protected will work here.

felixarntz replied:
Fair point. 👍

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -77,7 +77,7 @@ class WP_User {
 	 * @var string[]
 	 */
-	public $roles = array();
+	protected $roles = array();
```
Same here.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -86,7 +86,7 @@ class WP_User {
 	 *             and boolean values represent whether the user has that capability.
 	 */
-	public $allcaps = array();
+	protected $allcaps = array();
```
Same here.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -722,6 +742,7 @@ public function add_cap( $cap, $grant = true ) {
 	 * @param string $cap Capability name.
 	 */
 	public function remove_cap( $cap ) {
+		$this->load_capablity_data();
```
This feedback is not directly related to this line, but it looks like you're not loading the capability data in the `remove_all_caps()` method below. This could lead to unexpected consequences I think. It would be safer to add it there too.

spacedmonkey replied:
Fixed.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -876,9 +899,7 @@ public function for_site( $site_id = '' ) {
 
 		$this->cap_key = $wpdb->get_blog_prefix( $this->site_id ) . 'capabilities';
 
-		$this->caps = $this->get_caps_data();
-
-		$this->get_role_caps();
```
I think removing this without replacing it with some other mechanism to "reset" the flag that capability and role data are loaded will lead to problems. For example, if capabilities and roles were already loaded and then someone calls `WP_User::for_site()` with another site ID, the `WP_User` object would still have the old capabilities and roles, not the ones for the site requested.

peterwilsoncc replied:
Indeed, this appears to be the cause of the test failures for MS.

spacedmonkey replied:
Fixed.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -876,9 +899,7 @@ public function for_site( $site_id = '' ) {
-		$this->caps = $this->get_caps_data();
-
-		$this->get_role_caps();
+		wp_lazyload_user_meta( array( $this->ID ) );
```
I'm not very familiar with how metadata lazy-loading works, so I feel like I'm not the right person to review this part. For example, one potential caveat here is that user metadata is generally global, but some meta keys only apply to specific sites.

I wonder for instance why this method is called here. Since user metadata is global, why would it need to be loaded (again) when the site is switched? Shouldn't this call be somewhere else, or only be called if it hadn't been called before?

spacedmonkey replied:
This is the key to not loading user meta on every request and is required.

felixarntz replied:
I understand that this is useful, but I don't think it's required for this PR. This line fits perfectly into the scope of https://core.trac.wordpress.org/ticket/63021, but less into lazy-loading the role/cap properties of `WP_User`.

On another note, if we keep it, could you review my previous question please?
> Since user metadata is global, why would it need to be loaded (again) when the site is switched? Shouldn't this call be somewhere else, or only be called if it hadn't been called before?

felixarntz replied:
@spacedmonkey Can you please get back to me on this open question?

felixarntz reviewed (changes requested):
@spacedmonkey I think there is at least one bug with the code that needs to be addressed (related to switching to another site in Multisite).

Other than that, I think it would be great if you could separate the overall metadata lazyloader integration and handle that in another PR. The role and capability logic itself is expensive enough to be lazy-loaded, which is what most of this PR is about. That in itself is already complex and may have some quirks to consider, so I think separating the two concerns makes the PRs easier to review.

As an aside, I'm not very familiar with metadata lazy-loading, but I am familiar with how `WP_User` works, so I only feel comfortable with reviewing the latter.

spacedmonkey committed:
Hopefully fix multisite tests.

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey committed:
Force reload on remove_all_caps.

spacedmonkey committed:
Improve test.

spacedmonkey commented on src/wp-includes/user.php:
```diff
@@ -1216,6 +1216,21 @@ function get_user_meta( $user_id, $key = '', $single = false ) {
+/**
+ * Queue user meta for lazy-loading.
+ *
+ * @since 6.4.0
```
```suggestion
 * @since 6.8.0
```

spacedmonkey committed:
Update src/wp-includes/user.php

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -909,4 +938,18 @@ private function get_caps_data() {
 	}
+
+	/**
+	 * Load capability data.
```
```suggestion
	 * Loads capability data if it has not been loaded yet.
```

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -740,7 +767,8 @@ public function remove_cap( $cap ) {
 	public function remove_all_caps() {
 		global $wpdb;
-		$this->caps = array();
+		$this->caps        = array();
+		$this->loaded_caps = false;
 		delete_user_meta( $this->ID, $this->cap_key );
 		delete_user_meta( $this->ID, $wpdb->get_blog_prefix() . 'user_level' );
 		$this->get_role_caps();
```
This implementation looks a bit inconsistent with that from the other methods, and it is doing manually what `load_capability_data()` does. So I think it makes sense to use it:

```suggestion
		delete_user_meta( $this->ID, $this->cap_key );
		delete_user_meta( $this->ID, $wpdb->get_blog_prefix() . 'user_level' );
		$this->loaded_caps = false;
		$this->load_capability_data();
```

By calling `load_capability_data()` here, `$this->caps` will be automatically set to the empty array (if deleting the user metadata succeeded), and then `$this->get_role_caps()` will be called.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -513,6 +531,10 @@ public function get_role_caps() {
 
 		$wp_roles = wp_roles();
 
+		if ( ! $this->loaded_caps ) {
+			$this->caps = $this->get_caps_data();
+		}
+
```
This is unnecessary I think. The `WP_User` class has always populated `$this->caps` before, and not doing so anymore would be a breaking change. So we don't need this here, since `$this->caps` will always be correct prior to calling this method.
```suggestion
```

spacedmonkey replied:
The lines below checks if caps is an array. 

```
if ( is_array( $this->caps ) ) 
```

I wanted to ensure `$this->caps` was set correctly before doing this check.

felixarntz replied:
My point is that that's unnecessary to check, because it'll always be set correctly based on `$this->caps = $this->get_caps_data()`, which is called before this `get_role_caps()` method everywhere.

spacedmonkey replied:
The method `get_role_caps` is public. What you create a WP_User instance and the call `get_role_caps`?  It feels safer to just make sure that caps are loaded here. What does it hurt to have this here?

felixarntz replied:
Hmm, that's a fair point. Could you add an inline comment to clarify this? For example:
```suggestion
		// Edge-case: In case someone calls this method before lazy initialization, we need to initialize on demand.
		if ( ! $this->loaded_caps ) {
			$this->caps = $this->get_caps_data();
		}

```

felixarntz replied:
In addition, ideally this method would also set `$this->loaded_caps = true` in that situation: If someone calls `get_role_caps()` before initialization, the data is all set up the same way as `load_capability_data()` does it. So there would be no point of lazy-loading that data again afterwards.

felixarntz reviewed (commented):
@spacedmonkey The `WP_User` changes look almost good to commit, just a few points of minor feedback.

As mentioned, I would like for someone else that is more experienced with `WP_Metadata_Lazyloader` to review that portion of the PR.

peterwilsoncc commented on src/wp-includes/class-wp-user.php:
For IDEs, the now protected properties will need to be added to the list of properties available in the class level docblock. https://github.com/WordPress/wordpress-develop/blob/64cefd6a3d6f5f3e1ea859915b1f1d07f0519ef9/src/wp-includes/class-wp-user.php#L10-L38

spacedmonkey replied:
Fixed

peterwilsoncc commented on src/wp-includes/class-wp-user.php:
```diff
@@ -909,4 +938,18 @@ private function get_caps_data() {
+	 *
+	 * @since 6.8.0
+	 */
+	private function load_capablity_data() {
```
```suggestion
	private function load_capability_data() {
```

peterwilsoncc commented on tests/phpunit/tests/user/query.php:
```diff
@@ -159,16 +159,22 @@ public function test_get_all() {
 	 * @ticket 55594
 	 */
 	public function test_get_all_primed_users() {
+		$this->reset_lazyload_queue();
```
Do you know why this is needed, shouldn't the queue be clear?

peterwilsoncc reviewed (commented):
I've added a couple of notes and a question inline.

The change to cache_users, does make the change so user meta data is always lazy loaded and there doesn't appear to be a way for developers to trigger the active loading of user meta. 

Based on manual testing it appears pretty good.

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey committed:
Feedback from @peterwilsoncc & @felixarntz

felixarntz reviewed (commented):
@spacedmonkey @peterwilsoncc I am realizing just now that the change of making the `$caps`, `$roles`, and `$allcaps` properties protected may be considered a breaking change, albeit only relevant for edge cases.

They're still readable via magic getter and isset-er, but in theory a plugin doing some very niche capability modification stuff may be _writing_ these properties too. With the current state of this PR, that wouldn't work anymore.

I'm not overly concerned about this, since doing so would be a bad practice, but I still wanted to flag it. I think we have two options here:
* Either accept this as a edge-case BC break we're willing to make, punt this to 6.9 to land `early` in the milestone, and cover it in a dev note.
* Or add conditions for these to the `__set()` method to maintain backward compatibility (which is definitely more forgiving, but may continue encouraging this problematic pattern).

WDYT?

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey committed:
Remove lazy loading user meta.

spacedmonkey committed:
Remove unused wp_lazyload_user_meta function

The wp_lazyload_user_meta function was redundant and has been removed to simplify the codebase. This change helps reduce unnecessary code maintenance and improves overall clarity.

spacedmonkey committed:
Add unit tests for user role and capability handling

Introduce tests for `get_role_caps()` and lazy evaluation of user roles, capabilities, and allcaps. These tests ensure proper structure and expected values for user role-related data. Validates specific capabilities for contributors.

spacedmonkey reviewed (commented):
@peterwilsoncc @felixarntz @joemcgill As requested, I have removed all the lazy loading user meta stuff. This makes this patch much much smaller and easily to review. 

I have added some tests to valid that `get_role_caps` continues to work. Other existing tests cover add_cap, remove_cap, remove_all_caps, has_cap and set_role still work.

spacedmonkey commented:
> I am realizing just now that the change of making the `$caps`, `$roles`, and `$allcaps` properties protected may be considered a breaking change, albeit only relevant for edge cases.
> 
> They're still readable via magic getter and isset-er, but in theory a plugin doing some very niche capability modification stuff may be _writing_ these properties too. With the current state of this PR, that wouldn't work anymore.
> 
> I'm not overly concerned about this, since doing so would be a bad practice, but I still wanted to flag it. I think we have two options here:
> 
> * Either accept this as a edge-case BC break we're willing to make, punt this to 6.9 to land `early` in the milestone, and cover it in a dev note.
> * Or add conditions for these to the `__set()` method to maintain backward compatibility (which is definitely more forgiving, but may continue encouraging this problematic pattern).
> 
> WDYT?

If you are change the properties manaully, I feel like that are a pretty dangerous way of changing a user role. There are other filters and ways of doing this that are less problematic. I do believe this needs a detail dev note, but I feel like we should not enable this edge.

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -36,6 +36,9 @@
  * @property string $rich_editing
  * @property string $syntax_highlighting
  * @property string $use_ssl
+ * @property array $caps
+ * @property array $roles
+ * @property array $allcaps
```
Let's go with the more specific type notation, as it's already used in other places in this class. Also noting that the `bool[]` annotation below is actually incorrect, so we should fix that.
```suggestion
 * @property array<string, bool> $caps
 * @property string[] $roles
 * @property array<string, bool> $allcaps
```

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -62,7 +65,7 @@ class WP_User {
 	 * @var bool[] Array of key/value pairs where keys represent a capability name
 	 *             and boolean values represent whether the user has that capability.
```
See above. Not a problem in this PR, but since we're touching it, it would be good to fix this incorrect type annotation.
```suggestion
	 * @var array<string, bool> Array of key/value pairs where keys represent a capability name
	 *                          and boolean values represent whether the user has that capability.
```

felixarntz commented on src/wp-includes/class-wp-user.php:
```diff
@@ -87,7 +90,7 @@ class WP_User {
 	 * @var bool[] Array of key/value pairs where keys represent a capability name
 	 *             and boolean values represent whether the user has that capability.
```
See above:
```suggestion
	 * @var array<string, bool> Array of key/value pairs where keys represent a capability name
	 *                          and boolean values represent whether the user has that capability.
```

felixarntz reviewed (commented):
@spacedmonkey This looks almost good to go.

Just a few minor points here, and could you please get back to my feedback in https://github.com/WordPress/wordpress-develop/pull/5098#discussion_r1941557785? Still waiting for a reply to that.

swissspidy commented:
@spacedmonkey FYI there are some merge conflicts plus some open feedback above from Felix (whom I accidentally requested another review from, sorry!)

spacedmonkey commented on src/wp-includes/class-wp-user.php:
```diff
@@ -105,6 +108,15 @@ class WP_User {
+	/**
+	 * Flag for if capability is loaded.
+	 *
+	 * @since 6.8.0
```
```suggestion
	 * @since 6.9.0
```

spacedmonkey commented on src/wp-includes/class-wp-user.php:
```diff
@@ -910,4 +939,18 @@ private function get_caps_data() {
+	/**
+	 * Loads capability data if it has not been loaded yet.
+	 *
+	 * @since 6.8.0
```
```suggestion
	 * @since 6.9.0
```

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Felix Arntz <flixos90@gmail.com>

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey committed:
Add assertions for subscriber role in multisite user tests

mukeshpanchal27 commented on tests/phpunit/tests/user/multisite.php:
```diff
@@ -369,12 +369,13 @@ public function test_add_user_to_blog_subscriber() {
 
 		switch_to_blog( $site_id );
 		$user = get_user_by( 'id', $user_id );
+        $this->assertContains( 'subscriber', $user->roles, 'User should have subscriber role' );
```
```suggestion
		$user = get_user_by( 'id', $user_id );
		$this->assertContains( 'subscriber', $user->roles, 'User should have subscriber role' );
```

mukeshpanchal27 commented on tests/phpunit/tests/user/multisite.php:
```diff
@@ -369,12 +369,13 @@ public function test_add_user_to_blog_subscriber() {
 		wpmu_delete_user( $user_id );
 
-		$this->assertContains( 'subscriber', $user->roles );
+        $this->assertContains( 'subscriber', $user->roles, 'User should still have subscriber role' );
```
```suggestion
		$this->assertContains( 'subscriber', $user->roles, 'User should still have subscriber role' );
```

mukeshpanchal27 commented on src/wp-includes/class-wp-user.php:
```diff
@@ -106,6 +109,15 @@ class WP_User {
 	private $site_id = 0;
 
+
+	/**
```
```suggestion
	/**
```

mukeshpanchal27 reviewed (commented):
Quick minor feedbacks

felixarntz reviewed (approved):
@spacedmonkey LGTM, other than the coding standards issues to resolve :)

westonruter commented on tests/phpunit/tests/user/query.php:
```diff
@@ -162,13 +162,18 @@ public function test_get_all_primed_users() {
 
+		$users = $query->get_results();
+		foreach ( $users as $user ) {
+			$user->roles;
```
I understand this is just to initiate loading the role data, but it looks like an error:

<img width="474" height="154" alt="Image" src="https://github.com/user-attachments/assets/884ac531-e15e-4542-88af-cff2eadf3b24" />

So I suggest adding an assertion to at least look at the value:

```suggestion
			$this->assertIsArray( $user->roles );
			foreach ( $user->roles as $role ) {
				$this->assertIsString( $role );
			}
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -515,6 +536,11 @@ public function get_role_caps() {
 
 		$wp_roles = wp_roles();
 
+		// Edge-case: In case someone calls this method before lazy initialization, we need to initialize on demand.
```
Nit, this is two words:

```suggestion
		// Edge case: In case someone calls this method before lazy initialization, we need to initialize on demand.
```

westonruter commented:
> Trac ticket: https://core.trac.wordpress.org/ticket/58001

Shouldn't this rather reference https://core.trac.wordpress.org/ticket/63021 ?

**Update:** No! 😅

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>
Co-authored-by: Weston Ruter <westonruter@gmail.com>

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load

spacedmonkey commented:
@westonruter @mukeshpanchal27 Feedback actioned, can you take a look.

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -288,6 +299,10 @@ public function __isset( $key ) {
 		}
 
+		if ( in_array( $key, array( 'caps', 'allcaps', 'roles' ), true ) ) {
+			return isset( $this->$key );
```
Shouldn't it always return `true` here since they will always be populated magically?

```suggestion
			return true;
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -515,6 +535,11 @@ public function get_role_caps() {
 		$wp_roles = wp_roles();
 
+		// Edge case: In case someone calls this method before lazy initialization, we need to initialize on demand.
+		if ( ! $this->loaded_caps ) {
+			$this->caps = $this->get_caps_data();
+		}
```
Per above:

```suggestion
		if ( ! isset( $this->caps ) ) {
			$this->caps = $this->get_caps_data();
		}
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -876,11 +908,8 @@ public function for_site( $site_id = 0 ) {
-
-		$this->get_role_caps();
+		$this->cap_key     = $wpdb->get_blog_prefix( $this->site_id ) . 'capabilities';
+		$this->loaded_caps = false;
```
Per above.

```suggestion
		$this->caps        = null;
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -911,4 +940,18 @@ private function get_caps_data() {
+		if ( $this->loaded_caps ) {
+			return;
+		}
+		$this->caps = $this->get_caps_data();
+		$this->get_role_caps();
+		$this->loaded_caps = true;
```
Per above.

```suggestion
		$this->caps = $this->get_caps_data();
		$this->get_role_caps();
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -911,4 +940,18 @@ private function get_caps_data() {
+	 * @since 6.9.0
+	 */
+	private function load_capability_data() {
+		if ( $this->loaded_caps ) {
```
Per above.

```suggestion
		if ( isset( $this->caps ) ) {
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -106,6 +109,14 @@ class WP_User {
 	 */
 	private $site_id = 0;
 
+	/**
+	 * Flag for if capability is loaded.
+	 *
+	 * @since 6.9.0
+	 * @var bool
+	 */
+	private $loaded_caps = false;
+
```
Minor point: Instead of introducing a new `$loaded_caps` member variable, the `$caps` member could be typed as `array<string, bool>|null` instead of `array<string, bool>`. It could be default `null` whenever wanting to check if whether the caps have been loaded, you can just check if `isset( $this->caps )`. This would eliminate having to look at two different member variables.

```suggestion
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -60,10 +63,10 @@ class WP_User {
-	 * @var bool[] Array of key/value pairs where keys represent a capability name
-	 *             and boolean values represent whether the user has that capability.
+	 * @var array<string, bool> Array of key/value pairs where keys represent a capability name
+	 *                          and boolean values represent whether the user has that capability.
 	 */
-	public $caps = array();
+	protected $caps = array();
```
```suggestion
	 * @var array<string, bool>|null Array of key/value pairs where keys represent a capability name
	 *                               and boolean values represent whether the user has that capability.
	 * 
	 *                               Null if the caps have not been loaded yet.
	 */
	protected $caps = null;
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -742,10 +772,10 @@ public function remove_cap( $cap ) {
 		delete_user_meta( $this->ID, $this->cap_key );
 		delete_user_meta( $this->ID, $wpdb->get_blog_prefix() . 'user_level' );
-		$this->get_role_caps();
+		$this->loaded_caps = false;
```
Per above.

```suggestion
		$this->caps = null;
```

spacedmonkey committed:
Refactor lazy loading of user capabilities and refine related tests

Adjust `WP_User` properties and logic for lazy loading of capabilities, roles, and allcaps. Replace `loaded_caps` flag with null `caps` to indicate uninitialized state. Update unit tests to validate manual addition of user properties.

spacedmonkey committed:
Reorder `$caps` reset in `remove_all_caps` for consistency.

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -60,10 +63,10 @@ class WP_User {
-	 * @var bool[] Array of key/value pairs where keys represent a capability name
-	 *             and boolean values represent whether the user has that capability.
+	 * @var array<string, bool>|null Array of key/value pairs where keys represent a capability name
+	 *                          and boolean values represent whether the user has that capability.
```
```suggestion
	 *                               and boolean values represent whether the user has that capability.
```

westonruter commented on src/wp-includes/class-wp-user.php:
```diff
@@ -515,6 +534,11 @@ public function get_role_caps() {
 
 		$wp_roles = wp_roles();
 
+		// Edge case: In case someone calls this method before lazy initialization, we need to initialize on demand.
+		if ( ! isset( $this->caps ) ) {
+			$this->caps = $this->get_caps_data();
+		}
```
Why wouldn't this also just do?

```suggestion
		$this->load_capability_data();
```

spacedmonkey replied:
This was originally considers. But `load_capability_data` calls `get_role_caps`, which calls `load_capability_data`. This will loop forever.

westonruter replied:
Ah, OK 😄 ♾️

spacedmonkey committed:
Ensure `WP_User` properties reset to null when unset, add tests

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@gmail.com>

spacedmonkey reviewed (commented):
@felixarntz @westonruter I feedback. I also fixed some edge case unsetting properties and setting properties. The current PR has new tests and is pretty clear, IMO.

spacedmonkey commented on src/wp-includes/class-wp-user.php:
```diff
@@ -37,6 +37,9 @@
  * @property string $rich_editing
  * @property string $syntax_highlighting
  * @property string $use_ssl
+ * @property array<string, bool> $caps
```
```suggestion
 * @property array<string, bool>|null $caps
```

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey commented on src/wp-includes/class-wp-user.php:
```diff
@@ -37,6 +37,9 @@
  * @property string $rich_editing
  * @property string $syntax_highlighting
  * @property string $use_ssl
+ * @property array<string, bool>|null $caps
```
```suggestion
 * @property array<string, bool> $caps
```

spacedmonkey committed:
Apply suggestions from code review

mukeshpanchal27 commented on tests/phpunit/tests/user/capabilities.php:
```diff
@@ -992,6 +992,37 @@ public function test_add_role() {
 
+	/**
+	 * Test adding capabilities, roles, and allcaps manually to a user.
+	 *
+	 * @ticket 58001
+	 * @dataProvider data_add_user_properties_manually
```
```suggestion
	 * @ticket 58001
	 *
	 * @dataProvider data_add_user_properties_manually
	 *
	 * @param string $property_name  The property name to set.
	 * @param array  $property_value The property value to set.
	 * @param bool   $check_null     Whether to check that the property is null after unsetting it.
```

spacedmonkey committed:
Update tests/phpunit/tests/user/capabilities.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

spacedmonkey committed:
Merge branch 'trunk' into try/lazy-load
